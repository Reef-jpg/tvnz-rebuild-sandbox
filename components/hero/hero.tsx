import { MouseEvent, MouseEventHandler, useState } from "react";

import { Button } from "../button/button";
import { Link } from "../link/link";
import styled from "styled-components";

const StyledDiv = styled.div`
  width: 100%;
  min-height: 380px;

  position: relative;
  cursor: pointer;

  /* image styling */
  .image-wrapper {
    z-index: 0;
    width: 100%;
    max-width: 100%;
    max-height: 100%;

    position: absolute;

    img {
      width: 100%;
      height: auto;
      max-width: 100%;
      max-height: 100%;
    }

    /* gradient fade on image */
    &::before {
      width: 100%;
      height: 100%;

      display: block;
      position: absolute;

      content: "";
      background: linear-gradient(
        to bottom,
        transparent,
        rgba(17, 17, 17, 0.25),
        rgba(17, 17, 17, 0.75),
        #111
      );
    }
  }

  /* content styling */
  .content {
    z-index: 1;
    padding: calc(1rem + 17%) 4.75rem 0 4.75rem; // 17% for image offset

    position: relative;

    /* text */
    .title {
      margin: 0;
      font-weight: 700;
      font-size: 3rem;
      line-height: 1.1;
    }
    .description {
      max-width: 50%;
      margin: 1rem 0 2rem 0;
      font-size: 1rem;
      font-weight: 300;
      line-height: 21px;
    }
    .sub-title {
      margin-bottom: 0.5rem;
      font-weight: 500;
      letter-spacing: 1px;
      line-height: 1.37rem;
      text-transform: uppercase;
    }

    /* buttons */
    .buttons {
      display: flex;

      .button {
        margin-right: 1rem;
      }
    }

    .sponsor {
      /* scaling to 0.5x as the API is always meant to provide a 2x resolution image for this spot (see PLAY-3837) */
      transform: scale(0.5) translate(50%, 50%);

      margin-right: 4.75rem;
      position: absolute;
      bottom: 0;
      right: 0;

      display: flex;
      flex-direction: column;

      .label {
        /* double everything to offset the 0.5x of parent wrapper */
        margin-bottom: calc(4px * 2);
        font-size: calc(14px * 2);
        color: #bfbfbf;
      }
      .link {
        display: flex;
        flex-direction: column;
      }
    }
  }
`;

export const Hero = ({
  title,
  description,
  image,
  imageAlternativeText = "Hero Image",
  subTitle,
  heroLink,
  buttonLink,
  sponsorLogo,
}: {
  title: string;
  description: string;
  image: string;
  imageAlternativeText?: string;
  subTitle?: string;
  heroLink: string;
  buttonLink: string;
  sponsorLogo?: { label?: string; link: string; image: string };
}) => {
  // state for my list button
  const [faved, setFaved] = useState<boolean>(false);

  // click handlers
  const heroLinkHandler: MouseEventHandler = (event: MouseEvent) => {
    // only follow link if one of these elements was clicked
    const selectorsToLink = ["content", "description", "title", "sub-title"];
    if (
      heroLink &&
      selectorsToLink.some((x) =>
        (event?.target as HTMLElement)?.classList.contains(x)
      )
    )
      window.location.assign(heroLink);
  };

  const myListHandler = () => {
    setFaved(!faved);
  };

  return (
    <StyledDiv className="hero" onClick={heroLinkHandler}>
      {/* hero splash image */}
      <div className="image-wrapper">
        <img src={image} alt={imageAlternativeText} />
      </div>

      {/* content section of hero */}
      <div className="content">
        {/* text */}
        {subTitle && <div className="sub-title">{subTitle}</div>}
        <h1 className="title">{title}</h1>
        <p className="description">{description}</p>

        {/* buttons */}
        <div className="buttons">
          <Link href={buttonLink}>
            <Button text="watch now" icon="/icon-play.svg" />
          </Link>

          <Button
            text="my list"
            icon={faved ? "/icon-mylist-tick.svg" : "/icon-mylist.svg"}
            variant="secondary"
            className={faved ? "in-my-list" : "not-in-my-list"}
            onClick={myListHandler}
          />
        </div>

        {/* sponsor ad */}
        {sponsorLogo && (
          <div className="sponsor">
            <Link
              href={sponsorLogo.link}
              target="_blank"
              rel="nofollow noopener"
            >
              <p className="label">
                {sponsorLogo.label ? sponsorLogo.label : "Proud Sponsor"}
              </p>
              <img src={sponsorLogo.image} alt="Sponsor Logo" />
            </Link>
          </div>
        )}
      </div>
    </StyledDiv>
  );
};
