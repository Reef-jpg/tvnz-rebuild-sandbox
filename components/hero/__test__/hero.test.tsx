import renderer, { act } from 'react-test-renderer';
import { Hero } from '../hero';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

describe("Hero Component", () => {
  test("should render", () => {
    const { container } = render(
      <Hero title="" description="" image="" heroLink="" buttonLink="" />
    );

    expect(container.querySelector(".hero")).toBeTruthy();
    expect(container.querySelector(".hero")).toBeVisible();
  });

  test("should have title and sub title", () => {
    const { getByText } = render(
      <Hero
        title="Test Hero Title"
        subTitle="Sub Test"
        description=""
        image=""
        heroLink=""
        buttonLink=""
      />
    );

    expect(getByText("Test Hero Title")).toBeInTheDocument();
    expect(getByText("Test Hero Title")).toBeVisible();
    expect(getByText("Sub Test")).toBeInTheDocument();
    expect(getByText("Sub Test")).toBeVisible();
  });

  test("should have description", () => {
    const { getByText } = render(
      <Hero
        title=""
        description="hero desc"
        image=""
        heroLink=""
        buttonLink=""
      />
    );

    expect(getByText("hero desc")).toBeInTheDocument();
    expect(getByText("hero desc")).toBeVisible();
  });

  test("should have image", () => {
    const { getByAltText } = render(
      <Hero
        title=""
        description=""
        image="/test.jpg"
        imageAlternativeText="test hero alt text"
        heroLink=""
        buttonLink=""
      />
    );

    expect(getByAltText("test hero alt text")).toBeInTheDocument();
    expect(getByAltText("test hero alt text")).toBeVisible();
  });

  test("should render buttons", () => {
    const { getByText } = render(
      <Hero
        title=""
        description=""
        image=""
        heroLink="/hero"
        buttonLink="/button"
      />
    );

    expect(getByText("watch now")).toBeVisible();
    expect(getByText("watch now").closest("a")).toHaveAttribute(
      "href",
      "/button"
    );
    expect(getByText("my list")).toBeVisible();
  });

  describe("Sponsor Logo", () => {
    test("should render sponsor logo", () => {
      const { getByText, getByAltText } = render(
        <Hero
          title=""
          description=""
          image=""
          heroLink="/hero"
          buttonLink="/button"
          sponsorLogo={{
            link: "/sponsor",
            image: "/test.png",
            label: "Test Sponsor",
          }}
        />
      );

      expect(getByText("Test Sponsor")).toBeInTheDocument();
      expect(getByText("Test Sponsor")).toBeVisible();
      expect(getByText("Test Sponsor").closest("a")).toHaveAttribute(
        "href",
        "/sponsor"
      );
      expect(getByAltText("Sponsor Logo")).toBeInTheDocument();
      expect(getByAltText("Sponsor Logo")).toBeVisible();
    });

    test("should render fallback text", () => {
      const { getByText } = render(
        <Hero
          title=""
          description=""
          image=""
          heroLink="/hero"
          buttonLink="/button"
          sponsorLogo={{
            link: "/sponsor",
            image: "/test.png",
          }}
        />
      );

      expect(getByText("Proud Sponsor")).toBeInTheDocument();
      expect(getByText("Proud Sponsor")).toBeVisible();
    });
  });

  describe("Click Handlers", () => {
    test("click handler for my list should work", () => {
      const { getByText } = render(
        <Hero
          title="Test Hero State"
          description=""
          image=""
          heroLink=""
          buttonLink=""
        />
      );

      expect(getByText("my list").closest("button")).toHaveClass(
        "not-in-my-list"
      );

      act(() => {
        getByText("my list").click();
      });

      expect(getByText("my list").closest("button")).toHaveClass("in-my-list");
    });

    test("click handler for hero should work", () => {
      // setup
      const { location } = window;
      // @ts-ignore
      delete window.location;
      window.location = { ...location, assign: jest.fn() };

      const { getByText, container } = render(
        <Hero
          title="Title"
          subTitle="Sub Title"
          description="Desc"
          image=""
          heroLink="/hero"
          buttonLink="/button"
          sponsorLogo={{
            link: "/sponsor",
            image: "/test.png",
            label: "Test Sponsor",
          }}
        />
      );

      act(() => {
        // things that should trigger event
        getByText("Title").click();
        getByText("Desc").click();
        getByText("Sub Title").click();
        (container.querySelector(".content") as HTMLElement)?.click();

        // things that should not trigger the event
        getByText("watch now").click();
        getByText("my list").click();
        getByText("Test Sponsor").click();
      });

      expect(window.location.assign).toHaveBeenCalledTimes(4);
      expect(window.location.assign).toHaveBeenCalledWith("/hero");

      // cleanup
      window.location = location;
    });
  });

  test("should match snapshots", () => {
    const withProps = renderer
      .create(
        <Hero
          title="Hero Test"
          description="Description"
          image="/test-bg.jpg"
          imageAlternativeText="Pic of a test"
          subTitle="TEST"
          heroLink="/hero"
          buttonLink="/button"
        />
      )
      .toJSON();

    expect(withProps).toMatchSnapshot();
  });
});
