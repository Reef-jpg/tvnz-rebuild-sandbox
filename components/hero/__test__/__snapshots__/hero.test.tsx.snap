// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero Component should match snapshots 1`] = `
<div
  className="hero__StyledDiv-sc-s6ikjf-0 dKJFNe hero"
  onClick={[Function]}
>
  <div
    className="image-wrapper"
  >
    <img
      alt="Pic of a test"
      src="/test-bg.jpg"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="sub-title"
    >
      TEST
    </div>
    <h1
      className="title"
    >
      Hero Test
    </h1>
    <p
      className="description"
    >
      Description
    </p>
    <div
      className="buttons"
    >
      <a
        className="link__StyledLink-sc-k5dztf-0 jZDYYg link "
        href="/button"
      >
        <button
          className="button__StyledButton-sc-1476jkj-0 iOnKIy button  
        primary "
        >
          <span
            className="icon"
          >
            <img
              alt="Button Icon"
              src="/icon-play.svg"
            />
          </span>
          <span
            className="text"
          >
            watch now
          </span>
        </button>
      </a>
      <button
        className="button__StyledButton-sc-1476jkj-0 iOnKIy button  
        secondary not-in-my-list"
        onClick={[Function]}
      >
        <span
          className="icon"
        >
          <img
            alt="Button Icon"
            src="/icon-mylist.svg"
          />
        </span>
        <span
          className="text"
        >
          my list
        </span>
      </button>
    </div>
  </div>
</div>
`;
